crates/rpc/engine/mod.rs

```diff
@@ -102,22 +103,25 @@ pub fn new_payload_v3(
         return Ok(PayloadStatus::syncing());
     }
 
-    // Fetch parent block header and validate current header
-    if let Some(parent_header) = storage.get_block_header(block.header.number.saturating_sub(1))? {
-        if !validate_block_header(&block.header, &parent_header) {
-            return Ok(PayloadStatus::invalid_with_hash(
-                parent_header.compute_block_hash(),
-            ));
-        }
-    } else {
-        return Ok(PayloadStatus::syncing());
-    }
+    let latest_valid_hash = latest_valid_hash(&storage).map_err(|_| RpcErr::Internal)?;
 
     // Execute and store the block
     info!("Executing payload with block hash: {block_hash}");
-    execute_block(&block, &mut evm_state(storage.clone()))?;
-    info!("Block with hash {block_hash} executed succesfully");
-    info!("Block with hash {block_hash} added to storage");
+    match add_block(&block, storage.clone()) {
+        Err(ChainError::NonCanonicalBlock) => Ok(PayloadStatus::syncing()),
+        Err(ChainError::ParentNotFound) => Ok(PayloadStatus::invalid_with_err(
+            "Could not reference parent block with parent_hash",
+        )),
```

Comment:
Is this logic correct? Blocks without parents should lead to a sync to find the missing parent, blocks with lower block numbers than the canonical heads should lead to potential reorgs, etc

Reply:
You are correct, but since right now we don't support neither syncs nor reorgs I wasn't sure about how to solve it. - - NonCanonicalBlock is returned when the block is not the block number+1.
- ParentNotFound is returned when the parent block cannot by found by the parent_hash value in the new block's header.
How could we improve these error types? Maybe something like ShouldSync ShouldReorg to indicate that's what we should do but we cannot support it? 
My idea was to, for now, only allow adding block which can extend the canonical chain directly, that is only adding blocks which are children of the head of the chain.